js/parts-map/HeatmapSeries.js

```diff
@@ -179,32 +184,122 @@ seriesType('heatmap', 'scatter',
-            point.shapeType = 'rect';
-            point.shapeArgs = {
-                x: Math.min(x1, x2) + pointPadding,
-                y: Math.min(y1, y2) + pointPadding,
-                width: Math.max(Math.abs(x2 - x1) - pointPadding * 2, 0),
-                height: Math.max(Math.abs(y2 - y1) - pointPadding * 2, 0)
+            var pointAttr, dif, hasImage = (point.marker && point.marker.symbol || shape || '')
+                .match(/url/), cellAttr = point.getCellAttributes(), shapeArgs = {
+                x: Math.min(cellAttr.x1, cellAttr.x2),
+                y: Math.min(cellAttr.y1, cellAttr.y2),
+                width: Math.max(Math.abs(cellAttr.x2 - cellAttr.x1), 0),
+                height: Math.max(Math.abs(cellAttr.y2 - cellAttr.y1), 0)
             };
+            // If marker shape is regular (symetric), find shorter
+            // cell's side.
+            if (hasRegularShape) {
+                dif = Math.abs(shapeArgs.width - shapeArgs.height);
+                shapeArgs.x = Math.min(cellAttr.x1, cellAttr.x2) +
+                    (shapeArgs.width < shapeArgs.height ? 0 : dif / 2);
+                shapeArgs.y = Math.min(cellAttr.y1, cellAttr.y2) +
+                    (shapeArgs.width < shapeArgs.height ? dif / 2 : 0);
+                shapeArgs.width = shapeArgs.height =
+                    Math.min(shapeArgs.width, shapeArgs.height);
+            }
+            pointAttr = {
+                plotX: (cellAttr.x1 + cellAttr.x2) / 2,
+                plotY: (cellAttr.y1 + cellAttr.y2) / 2,
+                clientX: (cellAttr.x1 + cellAttr.x2) / 2,
+                shapeType: 'path',
+                shapeArgs: merge(true, shapeArgs, {
+                    d: symbols[shape](shapeArgs.x, shapeArgs.y, shapeArgs.width, shapeArgs.height)
+                })
+            };
+            if (hasImage) {
+                point.marker = {
+                    width: shapeArgs.width,
+                    height: shapeArgs.height
+                };
+            }
+            extend(point, pointAttr);
         });
         fireEvent(series, 'afterTranslate');
     },
+    /**
+     * @private
+     * @function Highcharts.seriesTypes.heatmap#pointAttribs
+     * @param {Highcharts.HeatmapPoint} point
+     * @param {string} state
+     * @return {Highcharts.SVGAttributes}
+     */
+    pointAttribs: function (point, state) {
+        var attr = seriesTypes.scatter.prototype.pointAttribs
+            .call(this, point, state), seriesOptions = this.options || {}, stateOptions, brightness;
+        // Apply lineColor, or set it to default series color.
+        attr.stroke = (point.marker && point.marker.lineColor ||
+            seriesOptions.marker && seriesOptions.marker.lineColor ||
+            this.color);
+        if (state) {
+            stateOptions = merge(seriesOptions.states[state], (seriesOptions.marker &&
+                seriesOptions.marker.states)[state], point.options.states &&
```

Comment:
What in case when `seriesOptions.marker.states` is not defined?

Reply:
They're always inherited from parent series `defaultOptions`